internal/deviations/deviations.go

```diff
@@ -47,13 +52,14 @@ import "flag"
 // Vendor deviation flags.
 var (
 	InterfaceEnabled = flag.Bool("deviation_interface_enabled", false,
-		"Device requires interface enabled leaf booleans to be explicitly set to true.")
+		"Device requires interface enabled leaf booleans to be explicitly set to true.  Fully-compliant devices should pass both with and without this deviation.")
```

Comment:
I think adding this additional description can cause confusion.  Passing without deviations is required for compliance.  Passing with a deviation is not compliant.  If the deviation is enabled, one cannot tell if the device is compliant.  The deviation must be disabled to know if the device is compliant.

Reply:
As discussed yesterday, the deviation itself should only be implemented using fully-compliant mechanisms.

* Device A: cannot set the subinterface enabled leaf. (not compliant)
* Device B: requires setting the interface and subinterface enabled leaf to true. (not compliant)
* Device C: interface and subinterface are both enabled by default (compliant)

| Device | Without Deviation | With Deviation |
| ---- | ---- | ---- |
| Device A | Pass | Fail (can't set leaf) |
| Device B | Fail (interface and subinterface are disabled) | Pass |
| Device C | Pass | Pass |

Again, you would favor Device A as compliant even though it is not, and disqualify Device C even though it is actually fully-compliant.

Reply:
@liulk  concerning `deviation_interface_enabled`, I think we need to treat ipv4/enabled and ipv6/enabled differently here.  I do not fully understand why a device that does not support setting ipv4/enabled (Device A) is not compliant, if it supports ipv4/enabled by default. 

I understand this is a deviation with OC, but that would be great if we focus on use cases for FP deviations. Do we have a use case that this leafs need to be used when device (i.e., Device C) enable ipv4 by default?

Reply:
I originally considered introducing a --deviation_subinterface_enabled in a follow up PR, but it sounds like you want separate --deviation_subinterface_ipv4_enabled and --deviation_subinterface_ipv6_enabled? What is the rationale for different IPv4 and IPv6 enabled behaviors?

Here is the OC verbiage, and this is the criteria I used for determining OC compliance.

* `/interfaces/interface/config/enabled` (default true) - This leaf contains the configured, desired state of the interface. Systems that implement the IF-MIB use the value of this leaf in the 'running' datastore to set IF-MIB.ifAdminStatus to 'up' or 'down' after an ifEntry has been initialized, as described in RFC 2863. Changes in this leaf in the 'running' datastore are reflected in ifAdminStatus, but if ifAdminStatus is changed over SNMP, this leaf is not affected.
* `/interfaces/interface/subinterfaces/subinterface/config/enabled` (default true) - This leaf contains the configured, desired state of the interface. Systems that implement the IF-MIB use the value of this leaf in the 'running' datastore to set IF-MIB.ifAdminStatus to 'up' or 'down' after an ifEntry has been initialized, as described in RFC 2863. Changes in this leaf in the 'running' datastore are reflected in ifAdminStatus, but if ifAdminStatus is changed over SNMP, this leaf is not affected.
* `/interfaces/interface/subinterfaces/subinterface/ipv4/config/enabled` (default true) - Controls whether IPv4 is enabled or disabled on this interface. When IPv4 is enabled, this interface is connected to an IPv4 stack, and the interface can send and receive IPv4 packets.
* `/interfaces/interface/subinterfaces/subinterface/ipv6/config/enabled` (default true) - Controls whether IPv6 is enabled or disabled on this interface. When IPv6 is enabled, this interface is connected to an IPv6 stack, and the interface can send and receive IPv6 packets.

There is also the "unnumbered" leaves which allows the L3 stack to be turned on without an L3 address. Currently, all our tests assign an IP address.

* `/interfaces/interface/subinterfaces/subinterface/ipv4/unnumbered/config/enabled` (default false) - Indicates that the subinterface is unnumbered. By default the subinterface is numbered, i.e., expected to have an IP address configuration.
* `/interfaces/interface/subinterfaces/subinterface/ipv6/unnumbered/config/enabled` (default false) - Indicates that the subinterface is unnumbered. By default the subinterface is numbered, i.e., expected to have an IP address configuration.

This means:

* Device A: cannot set the subinterface enabled leaf.
    * Not compliant because it doesn't support a path that is clearly defined in OC.
* Device B: requires setting the interface and subinterface enabled leaf to true.
    * Not compliant because OC default is enabled true, so the test should not have to set them.
* Device C: interface and subinterface are both enabled by default.
    * Compliant according to OC.

Now, I understand that Darren has a plan to adjust the OC verbiage to remove the ipv4/enabled and ipv6/enabled leaves. That is fine, and that will make both Device A and Device C compliant, and Device B will fail both with and without deviations because the test will not be able to set subinterface enabled, which becomes non-compliant under new OC verbiage.

What I want to establish here is the deviations guideline so that it adheres to the OC verbiage. This is not a good forum to discussing the merits of the OC verbiage itself.

Does that make sense?

Reply:
Thanks for the clarification!

Reply:
We may have different definitions for compliance.  IMO, the definition of compliance for featureprofiles test is the DUT is compliant to the requirements of the test.  This is a useful definition I think for the following reasons:

* An operator wants to be assured that a specific configuration will yield the exact same behavior on all devices which pass the test.
* OpenConfig and network devices in general may provide more than one way to achieve a desired behavior.

If a test passes with a deviation enabled, the operator needs to know the behavior requirement in the test is not being met.   The configuration the operating is generating is either not fully supported or will have some different behavior (deviation) than what the test is requiring.  This is why I think it is useful to defining passing with a deviation as not compliant with the test.  

I do agree that we can require that deviations only implement openconfig.

Reply:
Just for posterity to summarize my part of our discussion: my definition of compliance targets the OpenConfg public model rather than the test, and that's because:

 1. Tests should not use behavior outside of OpenConfig even when implementing a deviation.
 2. Tests are not written in a rigorous language that could be used to standardize OpenConfig behavior.

If we make tests be the compliance target, it can contradict the language in the OpenConfig public model and become a "de facto" standard even though they do not go through the same rigorous standardization process. But our definitions of compliance will converge when we agree that all tests, with or without deviations, only depend on OpenConfig behavior.

Tests typically only test a subset of OpenConfig, so they are already more lenient and represent the "low bar" of OpenConfig compliance. Deviation can make them even more lenient, and that's why the bar is even lower. Device should meet the higher bar without the deviation.

Reply:
To comment quickly on this part:

> it sounds like you want separate --deviation_subinterface_ipv4_enabled and --deviation_subinterface_ipv6_enabled? What is the rationale for different IPv4 and IPv6 enabled behaviors?

I believe my opinion here is a bit outside of the scope of the feature profile itself as it pertains to the merits of the ipv4 enabled leaf, but I would say that most implementations have a mechanism for having 'ipv6 enabled' which turns on ipv6 neighbor discovery and link-local addressing. In IPv6 it is possible for two hosts to communicate by simply 'turning on' ipv6, which is not the case in ipv4. This is a fundamental difference between ipv4 and ipv6, and can result in a difference in vendor behavior when doing mapping to openconfig as the meaning of ipv4 enabled is considerably more ambiguous than the the meaning of ipv6 enabled.

Reply:
Thanks for the clarification. Our tests using this deviation always assigns a global address, so this deviation flag is not meant to test unnumbered stack turnup.

We don't have a test plan written for the unnumbered stack behavior. In the future if someone adds a test for that, I think the test will additionally have to set `ipv6/unnumbered/config/enabled` to true per OC spec. The test will probably be very specific about which "enabled" combination to use, so it will not use a deviation.